code/game/machinery/_machines_base/machinery_damage.dm

```diff
@@ -0,0 +1,77 @@
+/obj/machinery/proc/take_damage(amount, damtype, silent = TRUE)
+	//Let's not bother initializing all the components for nothing
+	if(amount <= 0)
+		return
+	if(damtype != BRUTE && damtype != BURN && damtype != ELECTROCUTE)
+		return
+	if(!silent)
+		var/hitsound = 'sound/effects/hit_on_shattered_glass.ogg'
+		if(damtype == ELECTROCUTE)
+			hitsound = "sparks"
+		else if(damtype == BURN)
+			hitsound = 'sound/items/Welder.ogg'
+		playsound(src, hitsound, 20)
+		
+	world << "[src] is taking [amount] of [damtype] damage"
+	// Shielding components (armor/fuses) take first hit
+	var/list/shielding = get_all_components_of_type(/obj/item/stock_parts/shielding)
+	world << "Shields: [json_encode(shielding)]"
+	for(var/obj/item/stock_parts/shielding/soak in shielding)
+		if(damtype in soak.protection_types)
+			amount -= soak.take_damage(amount, damtype)
+	if(amount <= 0)
+		return
+
+	// If some damage got past, next it's generic (non-circuitboard) components
+	world << "Getting meat"
+	var/obj/item/stock_parts/victim = get_damageable_component()
+	while(amount > 0 && victim)
+		world << "Got [victim]"
+		amount -= victim.take_damage(amount, damtype)
+		victim = get_damageable_component()
+	if(amount <= 0)
+		return
+
+	world << "Damaging circuitboard"
+	// And lastly hit the circuitboard
+	victim = get_component_of_type(/obj/item/stock_parts/circuitboard)
+	victim.take_damage(amount, damtype)
+	
+/obj/machinery/proc/get_damageable_component()
+	var/list/victims = component_parts.Copy()
+	victims = shuffle(victims)
+	for(var/obj/item/stock_parts/component in victims)
+		// Circuitboards are handled separately
+		if(istype(component, /obj/item/stock_parts/circuitboard))
+			continue
+		// Don't damage what can't be repaired
+		if(!(component.part_flags & PART_FLAG_HAND_REMOVE))
+			continue
+		if(component.is_functional())
+			return component
+	for(var/path in uncreated_component_parts)
+		if(uncreated_component_parts[path])
+			return force_init_component(path)
```

Comment:
It just needs first damageable one, and since uncreated ones are created right here and now, they usually have health to spare.
Just realized if I do flag for being damageable they'll need a check here, will add that

Reply:
done